File: games/pong/src/main/kotlin/com/neoutils/engine/games/pong/PongScene.kt
package com.neoutils.engine.games.pong

import com.neoutils.engine.input.Key
import com.neoutils.engine.math.Vec2
import com.neoutils.engine.render.Color
import com.neoutils.engine.render.Renderer
import com.neoutils.engine.scene.Node
import com.neoutils.engine.scene.Scene

class PongScene(
    defaultWidth: Float = 800f,
    defaultHeight: Float = 600f,
) : Scene() {

    val leftScore: Score = Score()
    val rightScore: Score = Score()

    val leftPaddle: Paddle = Paddle(
        playFieldHeight = defaultHeight,
        upKey = Key.W,
        downKey = Key.S,
    ).apply { name = "left" }

    val rightPaddle: Paddle = Paddle(
        playFieldHeight = defaultHeight,
        ai = true,
    ).apply { name = "right" }

    val ball: Ball = Ball(fieldCenter = Vec2(defaultWidth / 2f, defaultHeight / 2f)) { scorer ->
        when (scorer) {
            Goal.Side.Left -> leftScore.increment()
            Goal.Side.Right -> rightScore.increment()
        }
    }

    private val topWall = Wall(Vec2(defaultWidth, WALL_THICKNESS)).apply { name = "topWall" }
    private val bottomWall = Wall(Vec2(defaultWidth, WALL_THICKNESS)).apply { name = "bottomWall" }

    private val leftGoal = Goal(Goal.Side.Left, Vec2(GOAL_THICKNESS, defaultHeight)).apply {
        name = "leftGoal"
    }
    private val rightGoal = Goal(Goal.Side.Right, Vec2(GOAL_THICKNESS, defaultHeight)).apply {
        name = "rightGoal"
    }

    private val centerLine = CenterLine(x = defaultWidth / 2f, height = defaultHeight)

    init {
        name = "PongScene"
        addChild(centerLine)
        addChild(leftPaddle)
        addChild(rightPaddle)
        addChild(ball)
        addChild(topWall)
        addChild(bottomWall)
        addChild(leftGoal)
        addChild(rightGoal)
        addChild(leftScore)
        addChild(rightScore)

        rightPaddle.aiTargetY = { ball.transform.position.y + ball.size / 2f }
        layout(defaultWidth, defaultHeight)
    }

    override fun onResize(width: Float, height: Float) {
        layout(width, height)
    }

    private fun layout(width: Float, height: Float) {
        // Walls
        topWall.size = Vec2(width, WALL_THICKNESS)
        topWall.transform = topWall.transform.copy(position = Vec2(0f, 0f))
        bottomWall.size = Vec2(width, WALL_THICKNESS)
        bottomWall.transform = bottomWall.transform.copy(
            position = Vec2(0f, height - WALL_THICKNESS)
        )

        // Goals just past the visible play field
        leftGoal.size = Vec2(GOAL_THICKNESS, height)
        leftGoal.transform = leftGoal.transform.copy(position = Vec2(-GOAL_THICKNESS, 0f))
        rightGoal.size = Vec2(GOAL_THICKNESS, height)
        rightGoal.transform = rightGoal.transform.copy(position = Vec2(width, 0f))

        // Paddles
        leftPaddle.playFieldHeight = height
        leftPaddle.transform = leftPaddle.transform.copy(
            position = Vec2(PADDLE_MARGIN, height / 2f - Paddle.HEIGHT / 2f)
        )
        rightPaddle.playFieldHeight = height
        rightPaddle.transform = rightPaddle.transform.copy(
            position = Vec2(width - PADDLE_MARGIN - Paddle.WIDTH, height / 2f - Paddle.HEIGHT / 2f)
        )

        // Ball + center line
        ball.fieldCenter = Vec2(width / 2f, height / 2f)
        ball.reset(serveToward = if (ball.velocity.x >= 0f) 1f else -1f)
        centerLine.x = width / 2f
        centerLine.height = height

        // Scores: left of and right of the center line
        val scoreY = 24f
        val scoreOffset = 80f
        leftScore.transform = leftScore.transform.copy(
            position = Vec2(width / 2f - scoreOffset, scoreY)
        )
        rightScore.transform = rightScore.transform.copy(
            position = Vec2(width / 2f + scoreOffset / 2f, scoreY)
        )
    }

    companion object {
        const val WALL_THICKNESS: Float = 8f
        const val GOAL_THICKNESS: Float = 8f
        const val PADDLE_MARGIN: Float = 32f
    }
}

private class CenterLine(var x: Float, var height: Float) : Node() {
    override fun onRender(renderer: Renderer) {
        val dashHeight = 12f
        val gap = 8f
        val color = Color(1f, 1f, 1f, 0.3f)
        var y = 0f
        while (y < height) {
            renderer.drawRect(
                com.neoutils.engine.math.Rect(Vec2(x - 1f, y), Vec2(2f, dashHeight)),
                color,
                filled = true,
            )
            y += dashHeight + gap
        }
    }
}

File: games/pong/src/main/resources/pong/scripts/pong_scene.py
# extends Scene

# Layout / orchestration script for the Pong root. Owns the resize-driven
# placement of walls, goals, paddles, scoreboard and ball, and wires the
# `scored` signal from ball.py into the leftScore / rightScore instances.

WALL_THICKNESS: float = 8.0
GOAL_THICKNESS: float = 8.0
PADDLE_MARGIN: float = 32.0
PADDLE_WIDTH: float = 16.0
PADDLE_HEIGHT: float = 96.0
SCORE_Y: float = 24.0
SCORE_OFFSET: float = 80.0


def _ready(self):
    self._last_w = -1.0
    self._last_h = -1.0
    _wire_scoring(self)


def _process(self, dt):
    # Scene.onResize isn't part of the ScriptHost hook contract, so the
    # scene's current size is polled here. Scene.size is kept current by the
    # runtime via Scene.resize() so the check is cheap.
    width = self._node.width
    height = self._node.height
    if width != self._last_w or height != self._last_h:
        self._last_w = width
        self._last_h = height
        if width > 0.0 and height > 0.0:
            _layout(self, width, height)


def _wire_scoring(self):
    ball_node = self._node.findChild("Ball")
    left_score_node = self._node.findChild("leftScore")
    right_score_node = self._node.findChild("rightScore")
    if ball_node is None:
        return
    ball = script_of(ball_node)
    left_score = script_of(left_score_node) if left_score_node is not None else None
    right_score = script_of(right_score_node) if right_score_node is not None else None
    if ball is None:
        return

    # Closures capture the scoreboards by ref; the bound method is plain
    # Python and GraalPy SAM-converts it for Signal.connect.
    def _on_scored(side):
        if side == "Left" and left_score is not None:
            left_score.increment()
        elif side == "Right" and right_score is not None:
            right_score.increment()

    ball.scored.connect(_on_scored)


def _layout(self, width, height):
    root = self._node
    top_wall = root.findChild("topWall")
    bottom_wall = root.findChild("bottomWall")
    left_goal = root.findChild("leftGoal")
    right_goal = root.findChild("rightGoal")
    left_paddle = root.findChild("left")
    right_paddle = root.findChild("right")
    ball_node = root.findChild("Ball")
    left_score_node = root.findChild("leftScore")
    right_score_node = root.findChild("rightScore")

    if top_wall is not None:
        top_wall.size = Vec2(width, WALL_THICKNESS)
        top_wall.transform = Transform(
            Vec2(0.0, 0.0), top_wall.transform.scale, top_wall.transform.rotation
        )
    if bottom_wall is not None:
        bottom_wall.size = Vec2(width, WALL_THICKNESS)
        bottom_wall.transform = Transform(
            Vec2(0.0, height - WALL_THICKNESS),
            bottom_wall.transform.scale,
            bottom_wall.transform.rotation,
        )
    if left_goal is not None:
        left_goal.size = Vec2(GOAL_THICKNESS, height)
        left_goal.transform = Transform(
            Vec2(-GOAL_THICKNESS, 0.0),
            left_goal.transform.scale,
            left_goal.transform.rotation,
        )
    if right_goal is not None:
        right_goal.size = Vec2(GOAL_THICKNESS, height)
        right_goal.transform = Transform(
            Vec2(width, 0.0), right_goal.transform.scale, right_goal.transform.rotation
        )

    if left_paddle is not None:
        left_paddle.transform = Transform(
            Vec2(PADDLE_MARGIN, height / 2.0 - PADDLE_HEIGHT / 2.0),
            left_paddle.transform.scale,
            left_paddle.transform.rotation,
        )
    if right_paddle is not None:
        right_paddle.transform = Transform(
            Vec2(width - PADDLE_MARGIN - PADDLE_WIDTH, height / 2.0 - PADDLE_HEIGHT / 2.0),
            right_paddle.transform.scale,
            right_paddle.transform.rotation,
        )

    if ball_node is not None:
        ball = script_of(ball_node)
        if ball is not None:
            ball.fieldCenter = Vec2(width / 2.0, height / 2.0)
            if root.isLive:
                serve_toward = 1.0 if ball._velocity.x >= 0.0 else -1.0
                ball.reset(serve_toward)

    if left_score_node is not None:
        left_score_node.transform = Transform(
            Vec2(width / 2.0 - SCORE_OFFSET, SCORE_Y),
            left_score_node.transform.scale,
            left_score_node.transform.rotation,
        )
    if right_score_node is not None:
        right_score_node.transform = Transform(
            Vec2(width / 2.0 + SCORE_OFFSET / 2.0, SCORE_Y),
            right_score_node.transform.scale,
            right_score_node.transform.rotation,
        )
